core/embed/rust/src/ui/model_tt/layout.rs

```diff
@@ -259,6 +262,23 @@ where
     }
 }
 
+/*
+impl ParagraphSource for Obj
+where
+{
+    fn collect<'a>(&self, sink: &mut impl Extend<Paragraph<&'a str>>) {
+        let mut iter_buf = IterBuf::new();
+        let iter = Iter::try_from_obj_with_buf(*self, &mut iter_buf);
+        let iter = unwrap!(iter);
+        for obj in iter {
+            let text: StrBuffer = unwrap!(obj.try_into());
+            let text = text.as_ref();
+            sink.extend(Some(Paragraph::new(&theme::TEXT_NORMAL, text)))
+        }
+    }
+}
```

Comment:
I didn't manage to implement taking the paragraphs directly out of `Obj` because I couldn't figure out the lifetimes:
```
error[E0597]: `text` does not live long enough
   --> src/ui/model_tt/layout.rs:275:24
    |
269 |     fn collect<'a>(&self, sink: &mut impl Extend<Paragraph<&'a str>>) {
    |                -- lifetime `'a` defined here
...
275 |             let text = text.as_ref();
    |                        ^^^^^^^^^^^^^ borrowed value does not live long enough
276 |             sink.extend(Some(Paragraph::new(&theme::TEXT_NORMAL, text)))
    |             ------------------------------------------------------------ argument requires that `text` is borrowed for `'a`
277 |         }
278 |     }
    |     - `text` dropped here while still borrowed

```

If we decide not to implement the trait for `Obj` then `ParagraphSource` can be redefined to provide random-access (instead of sequential) which would probably allow us to get rid of the large temporary array.

Reply:
doing this for `Paragraph<StrBuffer>` should be fine, no? the whole point of `StrBuffer` is to steal reasonable lifetimes from under the GC

> redefined to provide random-access (instead of sequential)

we can get random access from uPy lists too. it risks exceptions, but so does using the iterator.

Reply:
what is the status of this? or should we merge as-is and create an issue for `ParagraphSource for Obj`?

Reply:
In #2531 I implemented `ParagraphSource for Vec<Paragraph<Obj>>` - it seemed more straightforward to me to determine the number of paragraphs and their styling at the beginning and only call to uPy to get the contents. If I'm measuring this correctly `ParagraphVecLong<Obj>` has 388 bytes in firmware.

Reply:
And `ParagraphVecLong<StrBuffer>` has 516, might save some memory by converting to `ParagraphVecLong<Obj>`. Quick & dirty way to measure this:

```rust
macro_rules! dump_size {
    ($x:expr $(,)?) => {
        #[allow(unknown_lints, eq_op)]
        const _: [(); 0 - $x as usize] = [];
    };
}
use core::mem;
dump_size!(mem::size_of::<ParagraphVecLong<Obj>>());
```